### PR TITLE
DBZ-7777 Improve processing speed of async engine processors which use List#get()

### PR DESCRIPTION
**Pull request by vjuranek**
ParallelSmtAndConvertConsumerProcessors:
    
Before the change:
    
        Benchmark                                        (processingOrder)  (recordCount)  (threadCount)  Mode  Cnt    Score   Error  Units
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              1    ss       295.353           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              2    ss       310.652           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              4    ss       305.956           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              8    ss       334.755           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000             16    ss       306.477           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              1    ss       257.661           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              2    ss        78.385           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              4    ss        75.899           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              8    ss        81.068           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000             16    ss        94.506           s/op
        DebeziumEnginePerf.processRecordsEmbeddedEngine                N/A         100000            N/A    ss         0.857           s/op
    
After the change:
    
        Benchmark                                        (processingOrder)  (recordCount)  (threadCount)  Mode  Cnt  Score   Error  Units
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              1    ss       0.738           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              2    ss       0.404           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              4    ss       0.524           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              8    ss       0.505           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000             16    ss       0.428           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              1    ss       0.685           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              2    ss       0.400           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              4    ss       0.357           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              8    ss       0.508           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000             16    ss       0.410           s/op
        DebeziumEnginePerf.processRecordsEmbeddedEngine                N/A         100000            N/A    ss       0.857           s/op

ParallelSmtAndConsumerProcessors:

Before the change:
    
        Benchmark                                        (processingOrder)  (recordCount)  (threadCount)  Mode  Cnt   Score   Error  Units
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              1    ss       88.006           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              2    ss       90.226           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              4    ss       83.129           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              8    ss       81.417           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000             16    ss       90.936           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              1    ss       83.645           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              2    ss       77.527           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              4    ss       88.495           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              8    ss       79.590           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000             16    ss       90.246           s/op
        DebeziumEnginePerf.processRecordsEmbeddedEngine                N/A         100000            N/A    ss        0.125           s/op
    
After the change:
    
        Benchmark                                        (processingOrder)  (recordCount)  (threadCount)  Mode  Cnt  Score   Error  Units
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              1    ss       0.514           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              2    ss       0.391           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              4    ss       0.467           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000              8    ss       0.393           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine               ORDERED         100000             16    ss       0.462           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              1    ss       0.410           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              2    ss       0.431           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              4    ss       0.460           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000              8    ss       0.419           s/op
        DebeziumEnginePerf.processRecordsAsyncEngine             UNORDERED         100000             16    ss       0.448           s/op
        DebeziumEnginePerf.processRecordsEmbeddedEngine                N/A         100000            N/A    ss       0.230           s/op

Here, as there is no SMT and also no converter involved, it's IMHO expected that async engine is slower because of thread management overhead.

### CODE REVIEW

**Comment by mfvitale** [comment]
Wow it's an huge improvement. Still don't know why there is this difference given that the `List` implementation here seems to be an `ArrayList`

https://stackoverflow.com/questions/1879255/performance-of-traditional-for-loop-vs-iterator-foreach-in-java

**Comment by vjuranek** [comment]
@mfvitale if you take a look on DBZ-7777 and screenshot from async profiler, it's actually not `ArrayList` what is slow, it's `LinkedList` which you get from `poll()` method. So basically what was terribly slow was `records.get(i)`. As as I'd need some counter to access futures in `ArrayList` when switched to `for (SourceRecord r : records)`,  I decided not to `List.get()` anywhere and switch to array. But yes, this probably wasn't needed (I didn't do any measurements with preserved `ArrayList` for futures).

**Comment by vjuranek** [comment]
now, reading SO answer, I guess their observation is in concordance with my observation here: 

> All the other five took less than 20 milliseconds to iterate over the whole list. 
> Using list.get(i) on a LinkedList 100,000 times took more than 2 minutes (!) to complete (60,000 times slower).

**Comment by mfvitale** [comment]
> it's `LinkedList` which you get from `poll()` method.

Yes, that's the problem. Great finding.

File: debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java [hunk 1]
```diff
@@ -40,14 +39,20 @@ public class ParallelSmtAndConvertAsyncConsumerProcessor<R> extends AbstractReco
     @Override
     public void processRecords(final List<SourceRecord> records) throws Exception {
         LOGGER.debug("Thread {} is submitting {} records for processing.", Thread.currentThread().getName(), records.size());
-        final List<Future<Void>> recordFutures = new ArrayList<>(records.size());
-        records.stream().forEachOrdered(
-                r -> recordFutures.add(recordService.submit(new ProcessingCallables.TransformConvertConsumeRecord<>(r, transformations, convertor, consumer))));
+
+        final Future<Void>[] recordFutures = new Future[records.size()];
+        int i = 0;
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java:44]: Given that the `i` is no longer local to `for` would it be possible to give it a more descriptive name here  and in other parts of the code too?

Comment (vjuranek) [reply debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java]: sure, changed to `arrayIndex`

File: debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java [hunk 2]
```diff
@@ -40,14 +39,20 @@ public class ParallelSmtAndConvertAsyncConsumerProcessor<R> extends AbstractReco
     @Override
     public void processRecords(final List<SourceRecord> records) throws Exception {
         LOGGER.debug("Thread {} is submitting {} records for processing.", Thread.currentThread().getName(), records.size());
-        final List<Future<Void>> recordFutures = new ArrayList<>(records.size());
-        records.stream().forEachOrdered(
-                r -> recordFutures.add(recordService.submit(new ProcessingCallables.TransformConvertConsumeRecord<>(r, transformations, convertor, consumer))));
+
+        final Future<Void>[] recordFutures = new Future[records.size()];
+        int arrayIndex = 0;
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java:44]: I was thinking about this alternative too
```java
        Iterator<SourceRecord> recordsIterator = records.iterator();
        for (int i = 0; recordsIterator.hasNext(); i++) {
            recordFutures[i] = recordService.submit(new ProcessingCallables.TransformConvertConsumeRecord<>(r, transformations, convertor, consumer));
        }
```
choose the one you like more

Comment (vjuranek) [reply debezium-embedded/src/main/java/io/debezium/embedded/async/ParallelSmtAndConvertAsyncConsumerProcessor.java:44]: I think this one is actually better - it's 2 lines shorter (per processor) and I'd say also more idiomatic, so let's use this one. Thanks!

**Comment by jpechane** [comment]
@vjuranek Applied, thanks!

**Pull request merged**